pydantic_ai_slim/pydantic_ai/_tool_manager.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+from collections.abc import Iterable
+from dataclasses import dataclass, replace
+from typing import Any, Generic
+
+from pydantic import ValidationError
+
+from pydantic_ai.output import DeferredToolCalls
+
+from . import messages as _messages
+from ._run_context import AgentDepsT, RunContext
+from .exceptions import ModelRetry, ToolRetryError, UnexpectedModelBehavior
+from .messages import ToolCallPart
+from .tools import ToolDefinition
+from .toolsets.abstract import AbstractToolset, ToolsetTool
+
+
+@dataclass
+class ToolManager(Generic[AgentDepsT]):
+    """TODO: Update docstring -- A toolset that caches the wrapped toolset's tool definitions for a specific run step and handles retries."""
+
+    ctx: RunContext[AgentDepsT]
+    toolset: AbstractToolset[AgentDepsT]
+    tools: dict[str, ToolsetTool[AgentDepsT]]
+
+    @classmethod
+    async def build(cls, toolset: AbstractToolset[AgentDepsT], ctx: RunContext[AgentDepsT]) -> ToolManager[AgentDepsT]:
+        return cls(
+            ctx=ctx,
+            toolset=toolset,
+            tools=await toolset.get_tools(ctx),
+        )
+
+    async def for_run_step(self, ctx: RunContext[AgentDepsT]) -> ToolManager[AgentDepsT]:
+        if ctx == self.ctx:
+            return self
+        else:
+            return await self.__class__.build(self.toolset, replace(ctx, retries=self.ctx.retries))
+
+    @property
+    def tool_defs(self) -> list[ToolDefinition]:
+        return [tool.tool_def for tool in self.tools.values()]
+
+    def get_tool_def(self, name: str) -> ToolDefinition | None:
+        """Get the tool definition for a given tool name, or `None` if the tool is unknown."""
+        try:
+            return self.tools[name].tool_def
+        except KeyError:
+            return None
+
+    async def handle_call(self, call: ToolCallPart, allow_partial: bool = False) -> Any:
```

Comment:
Have you implemented / planned how to implement instrumentation for this? This feels like a good candidate for otel instrumentation, maybe there's another place for it or whatever.